src/client/datascience/data-viewing/jupyterVariableDataProvider.ts

```diff
@@ -0,0 +1,104 @@
+
+@injectable()
+export class JupyterVariableDataProvider implements IJupyterVariableDataProvider {
+    private initialized: boolean = false;
+
+    constructor(
+        @inject(IJupyterVariables) @named(Identifiers.ALL_VARIABLES) private variableManager: IJupyterVariables,
+        @inject(DataViewerDependencyService) private dependencyService: DataViewerDependencyService,
+        private variable: IJupyterVariable,
+        private notebook: INotebook
+    ) {}
+
+    /**
+     * Normalizes column types to the types the UI component understands.
+     * Defaults to 'string'.
+     * @param columns
+     * @returns Array of columns with normalized type
+     */
+    private static getNormalizedColumns(columns: { key: string; type: string }[]): { key: string; type: ColumnType }[] {
+        return columns.map((column: { key: string; type: string }) => {
+            let normalizedType: ColumnType;
+            switch (column.type) {
+                case 'bool':
+                    normalizedType = ColumnType.Bool;
+                    break;
+                case 'integer':
+                case 'int32':
+                case 'int64':
+                case 'float':
+                case 'float32':
+                case 'float64':
+                case 'number':
+                    normalizedType = ColumnType.Number;
+                    break;
+                default:
+                    normalizedType = ColumnType.String;
+            }
+            return {
+                key: column.key,
+                type: normalizedType
+            };
+        });
+    }
+
+    public dispose(): void {
+        return;
+    }
+
+    public async getDataFrameInfo(): Promise<IDataFrameInfo> {
+        await this.ensureInitialized();
+        return {
+            columns: this.variable.columns
+                ? JupyterVariableDataProvider.getNormalizedColumns(this.variable.columns)
+                : this.variable.columns,
+            indexColumn: this.variable.indexColumn,
+            rowCount: this.variable.rowCount || 0
+        };
+    }
+
+    public async getAllRows() {
+        let allRows: JSONObject = {};
+        await this.ensureInitialized();
+        if (this.variable.rowCount) {
+            allRows = await this.variableManager.getDataFrameRows(
+                this.variable,
+                this.notebook,
+                0,
+                this.variable.rowCount
+            );
+        }
+        return allRows;
+    }
+
+    public async getRows(start: number, end: number) {
+        let rows: JSONObject = {};
+        await this.ensureInitialized();
+        if (this.variable.rowCount) {
+            rows = await this.variableManager.getDataFrameRows(this.variable, this.notebook, start, end);
+        }
+        return rows;
+    }
+
+    private async ensureInitialized(): Promise<void> {
+        // Postpone pre-req and variable initialization until data is requested.
+        if (!this.initialized) {
```

Comment:
```suggestion
        if (!this.initialized) {
        	this.initialized = true;
```

I think we might want to move `this.initialized = true` to immediately after the if condition.
Else what could happen is, if this method is called multiple times, then we end up calling `this.dependencyService.checkAndInstl....` multiple times.

Meaning we could have multiple prompts...

Reply:
Also the variable needs to be fetched just once.

Reply:
Good catch, thanks! :)

Reply:
Done. :)